### PR TITLE
Draw Waypoints as beacon beams

### PR DESCRIPTION
**Pull request by byBackfish**
Implement [waypoints on hud with beams](https://feedback.wynntils.com/posts/17/waypoints-on-hud-with-beams)

Draws a beacon beam on the location of each waypoint. Toggle-able

### CODE REVIEW

**Comment by ryanzhoudev** [comment]
Not a huge fan of this implementation. I feel like this is too basic to be useful in most situations.

The main issue I have with this is that you can't control what types or groups of waypoints are given a beacon. This can prove less than useful for those that record all chests, including T1 and T2. Honestly, even with just my T3/T4, it is quite overwhelming and waypoints with actual purpose (eg. gathering spot markers) lose meaning.

I would love to see a couple things such as
- Ability to select groups of waypoints instead of a global on/off switch (maybe a custom gui? with check boxes like the chat tab gui?)
- Option to have beacons display Y location data similarly to journeymap
- Option to only display beacons within a certain radius of the player
- Option to have certain groups or waypoints have their names visible

**Comment by byBackfish** [comment]
Actually, now that I think about it, adding a checkbox to each individual waypoint would be nice. This way you can make only the most important waypoints render a Beacon. And then making waypoints marked as "always show" (I forgot what it's actually called right now), always render, and the rest only rendering if the player is near. 
What are you thoughts on that?

**Comment by ryanzhoudev** [comment]
Sounds good, though I still think there should be a way to mark a batch/group of waypoints to show up with beacon

**Comment by byBackfish** [comment]
I hope i did not forget to add something of the list you made. If I did, please let me know

**Comment by byBackfish** [comment]
image

**Comment by byBackfish** [comment]
^ global beacon beam settings menu

**Comment by kristofbolyai** [comment]
> image
> 
> 

How does that top text look in smaller windows?

**Comment by byBackfish** [comment]
> > image
> 
> How does that top text look in smaller windows?

Good point. I split it up into two lines now, as long as you are playing on a reasonable resolution, you wont have any problems not being able to see the text fully

**Comment by ryanzhoudev** [comment]
image

Empty category in config book?

**Comment by ryanzhoudev** [comment]
The buttons here do a funny thing where they just decide to resize themselves

**Comment by ryanzhoudev** [comment]
Also, I think you should add some text letting the user know that by default, waypoints will only render when the player is close to them. I was kind of confused when I enabled everything and no beams appeared. Could be done under/near the existing text in the group selection screen?

**Comment by byBackfish** [comment]
Yep, this commit should fix/implement those things. Thanks for the feedback

**Comment by ryanzhoudev** [comment]
This feels decent in-game now. I will look at the code tomorrow

File: src/main/java/com/wynntils/modules/map/overlays/ui/WaypointCreationMenu.java [hunk 1]
```diff
@@ -291,6 +300,9 @@ protected void actionPerformed(GuiButton button) {
             setZoomNeeded(MapWaypointIcon.ANY_ZOOM);
         } else if (button == hiddenButton) {
             setZoomNeeded(MapWaypointIcon.HIDDEN_ZOOM);
+        } else if(button == toggleBeaconBeam) {
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/map/overlays/ui/WaypointCreationMenu.java]:
```suggestion
        } else if (button == toggleBeaconBeam) {
```

File: src/main/java/com/wynntils/modules/map/overlays/ui/WaypointCreationMenu.java [hunk 2]
```diff
@@ -99,6 +102,9 @@ public WaypointCreationMenu(WaypointProfile wp, GuiScreen previousGui) {
         buttonList.add(saveButton = new GuiButton(103, this.width/2 + 25, saveButtonHeight, 45, 18, "Save"));
         saveButton.enabled = false;
 
+        beaconBeam = wp != null && wp.shouldShowBeaconBeam();
+        buttonList.add(toggleBeaconBeam = new GuiButton(104, this.width/2 - 150, this.height/2 + 10, 18, 18, beaconBeam ? ChatFormatting.GREEN + "✔" : ChatFormatting.RED + "╳"));
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/map/overlays/ui/WaypointCreationMenu.java]:
```suggestion
        buttonList.add(toggleBeaconBeam = new GuiButton(104, this.width/2 - 150, this.height/2 + 10, 18, 18, beaconBeam ? ChatFormatting.GREEN + "✔" : ChatFormatting.RED + "X"));
```

I am pretty sure other X buttons in Wynntils are just a standard `X`

File: src/main/java/com/wynntils/modules/map/overlays/ui/WaypointBeaconGroupMenu.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  @Override
+  protected void actionPerformed(GuiButton b) throws IOException {
+    groupButtons.forEach((type, button) -> {
+      if (b == button) {
+        groupValues.put(type, !groupValues.getOrDefault(type, false));
+        button.displayString = getText(type);
+      }
+    });
+
+    if(b == doneButton) {
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/map/overlays/ui/WaypointBeaconGroupMenu.java]:
```suggestion
    if (b == doneButton) {
```

File: src/main/java/com/wynntils/modules/map/managers/NametagManager.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    GlStateManager.rotate(McIf.mc().getRenderManager().playerViewX, 1.0f, 0.0f, 0.0f);
+    GlStateManager.translate(0f, -0.25f, 0f);
+    GlStateManager.rotate(-McIf.mc().getRenderManager().playerViewX, 1.0f, 0.0f, 0.0f);
+    GlStateManager.rotate(McIf.mc().getRenderManager().playerViewY, 0.0f, 1.0f, 0.0f);
+    drawNametag(TextFormatting.YELLOW.toString() + Math.round(dist) + "m");
+    GlStateManager.popMatrix();
+    GlStateManager.disableLighting();
+  }
+
+  private static void drawNametag(String text) {
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/map/managers/NametagManager.java]: Would like to see some spacing in this method. Maybe even a comment. Just looking to improve readability.

**Review by magicus — APPROVED** [review]

File: src/main/java/com/wynntils/modules/map/instances/WaypointProfile.java [hunk 5]
```diff
@@ -27,8 +27,13 @@ public class WaypointProfile {
     WaypointType type;
     WaypointType group = null;
 
+    boolean showBeaconBeam;
+
+    public WaypointProfile(String name, double x, double y, double z, CustomColor color, WaypointType type, int zoomNeeded, boolean showBeaconBeam) {
+        this.name = name; this.x = x; this.y = y; this.z = z; this.color = color; this.type = type; this.zoomNeeded = zoomNeeded; this.showBeaconBeam = showBeaconBeam;
```
Comment (magicus) [inline src/main/java/com/wynntils/modules/map/instances/WaypointProfile.java]: The code quality tool complains about this, and I actully agree. You need one assignment per line. I realize this is not new code, but please fix it.

**Review by magicus — CHANGES_REQUESTED** [review]

**Review by magicus — DISMISSED** [review]

File: src/main/java/com/wynntils/modules/map/overlays/ui/WaypointBeaconGroupMenu.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      buttonList.add(button);
+      groupButtons.put(type, button);
+      currentY += 30;
+    }
+
+    doneButton = new GuiButton(0,this.width/2-fontRenderer.getStringWidth("Done"), this.height - 100, 45, 18, "Done");
+    buttonList.add(doneButton);
+  }
+
+  public String getText(WaypointProfile.WaypointType type){
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/map/overlays/ui/WaypointBeaconGroupMenu.java]:
```suggestion
  public String getText(WaypointProfile.WaypointType type) {
```

File: src/main/java/com/wynntils/modules/map/overlays/ui/WaypointBeaconGroupMenu.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+    for (WaypointProfile.WaypointType type : WaypointProfile.WaypointType.values()) {
+      GuiButton button = new GuiButton(0, currentX, currentY, fontRenderer.getStringWidth(getText(type)) + 25, 20, getText(type));
+      button.enabled = true;
+      buttonList.add(button);
+      groupButtons.put(type, button);
+      currentY += 30;
+    }
+
+    doneButton = new GuiButton(0,this.width/2-fontRenderer.getStringWidth("Done"), this.height - 100, 45, 18, "Done");
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/map/overlays/ui/WaypointBeaconGroupMenu.java]:
```suggestion
    doneButton = new GuiButton(0, this.width/2-fontRenderer.getStringWidth("Done"), this.height - 100, 45, 18, "Done");
```

File: src/main/java/com/wynntils/modules/map/overlays/ui/WaypointBeaconGroupMenu.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+  @Override
+  public void onRenderPreUIE(ScreenRenderer render) {}
+
+  @Override
+  public void onRenderPostUIE(ScreenRenderer render) {
+    int currentX = width / 2 - 160;
+    int currentY = 100;
+
+    for(WaypointProfile.WaypointType type : WaypointProfile.WaypointType.values()){
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/map/overlays/ui/WaypointBeaconGroupMenu.java]:
```suggestion
    for (WaypointProfile.WaypointType type : WaypointProfile.WaypointType.values()) {
```

File: src/main/java/com/wynntils/modules/map/overlays/ui/WaypointBeaconGroupMenu.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+      }
+    });
+
+    if (b == doneButton) {
+      save();
+      mc.displayGuiScreen(prev);
+    }
+  }
+
+  void save(){
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/map/overlays/ui/WaypointBeaconGroupMenu.java]:
```suggestion
  void save() {
```

**Review by ryanzhoudev — COMMENTED** [review]
I think this is good after these few spacing fixes.

**Review by ryanzhoudev — APPROVED** [review]

**Review by magicus — APPROVED** [review]

**Comment by Endermity** [comment]
Maybe this needs some new features?
https://feedback.wynntils.com/posts/145/add-option-to-change-the-render-distance-for-waypoint-beam

**Pull request merged**